logic/resolvable/Unifier.java

```diff
@@ -358,48 +359,33 @@ static Set<? extends ValueConstraint.Constant<?>> constantValueConstraints(Set<V
             return iterate(values).filter(ValueConstraint::isConstant).map(ValueConstraint::asConstant).toSet();
         }
 
-        static Function<com.vaticle.typedb.core.concept.thing.Attribute, Boolean> valuePredicate(ValueConstraint.Constant<?> value) {
-            Function<com.vaticle.typedb.core.concept.thing.Attribute, Boolean> predicateFn;
+        static <T> Function<com.vaticle.typedb.core.concept.thing.Attribute, Boolean> valuePredicate(ValueConstraint.Constant<T> value) {
             assert !value.isVariable();
-            if (value.predicate().isEquality()) {
-                PredicateOperator.Equality operator = PredicateOperator.Equality.of(value.predicate().asEquality());
-                if (value.isLong()) {
-                    predicateFn = (a) ->
-                            PredicateArgument.Value.LONG.apply(operator, ((AttributeImpl<?>) a).readableVertex(), value.asLong().value());
-                } else if (value.isDouble()) {
-                    predicateFn = (a) ->
-                            PredicateArgument.Value.DOUBLE.apply(operator, ((AttributeImpl<?>) a).readableVertex(), value.asDouble().value());
-                } else if (value.isBoolean()) {
-                    predicateFn = (a) ->
-                            PredicateArgument.Value.BOOLEAN.apply(operator, ((AttributeImpl<?>) a).readableVertex(), value.asBoolean().value());
```

Comment:
it's was strange that we had to reach in and apply a `PredicateArgument` operation directly. Now, we only access `Predicate` for normal comparisons, and apply `PredicateOperator` directly for special cases like regexes and `contains`.